File: packages/transport-mcp/ts/src/authenticate.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Transport-neutral MCP authentication: bearer verify, revocation check, typed result.

import {
  ScopeInsufficientError,
  TokenInvalidError,
  ZoneInvalidError,
  verify,
} from '@caracalai/identity'
import type { RevocationStore } from '@caracalai/revocation'
import type { AuthResult } from './types.js'

export interface AuthDeps {
  issuer: string
  audience: string
  zoneId?: string
  requiredScopes?: string[]
  revocations: RevocationStore
}

export function extractBearer(authHeader: string | undefined): string | null {
  if (!authHeader?.startsWith('Bearer ') || authHeader.length <= 7) return null
  const token = authHeader.slice(7).trim()
  return token === '' ? null : token
}

export async function authenticate(token: string, deps: AuthDeps): Promise<AuthResult> {
  if (!token) {
    return { ok: false, error: { code: 'missing_token', description: 'Missing bearer token' } }
  }

  try {
    const claims = await verify(token, {
      issuer: deps.issuer,
      audience: deps.audience,
      zoneId: deps.zoneId,
      requiredScopes: deps.requiredScopes,
    })
    if (claims.sid && (await deps.revocations.isRevoked(claims.sid))) {
      return { ok: false, error: { code: 'session_revoked', description: 'Session revoked' } }
    }
    return { ok: true, principal: claims }
  } catch (err) {
    if (err instanceof ScopeInsufficientError) {
      return { ok: false, error: { code: 'insufficient_scope', description: err.message } }
    }
    if (err instanceof ZoneInvalidError) {
      return { ok: false, error: { code: 'invalid_zone', description: 'Token zone validation failed' } }
    }
    if (err instanceof TokenInvalidError) {
      return { ok: false, error: { code: 'invalid_token', description: 'Token validation failed' } }
    }
    return { ok: false, error: { code: 'invalid_token', description: 'Token validation failed' } }
  }
}

File: packages/transport-mcp/python/caracalai_transport_mcp/authenticate.py
# Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
# Caracal, a product of Garudex Labs
#
# Transport-neutral MCP authentication: identity verify and revocation check.

from __future__ import annotations

from caracalai_identity import verify_token
from caracalai_revocation import RevocationStore

from .types import AuthError, AuthResult


def extract_bearer(auth_header: str | None) -> str | None:
    if not auth_header or not auth_header.startswith("Bearer ") or len(auth_header) <= 7:
        return None
    token = auth_header[7:].strip()
    return token or None


async def authenticate(
    token: str,
    issuer: str,
    audience: str,
    required_scopes: list[str] | None,
    expected_zone_id: str | None,
    revocations: RevocationStore,
) -> AuthResult:
    if not token:
        return AuthResult(None, AuthError("missing_token", "Missing bearer token"))

    try:
        claims = await verify_token(token, issuer, audience, required_scopes, expected_zone_id)
    except PermissionError as err:
        return AuthResult(None, AuthError("insufficient_scope", str(err)))
    except ValueError as err:
        message = str(err)
        if "zone" in message.lower():
            return AuthResult(None, AuthError("invalid_zone", "Token zone validation failed"))
        return AuthResult(None, AuthError("invalid_token", "Token validation failed"))

    sid = claims.get("sid")
    if isinstance(sid, str) and sid and revocations.is_revoked(sid):
        return AuthResult(None, AuthError("session_revoked", "Session revoked"))

    return AuthResult(claims, None)
